applications: nrf_desktop: Add settings loader documentation (#2047)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/settings_load.rst

.. _settings_load:

> Settings load module

Reviewer: I suggest to change the module name to `Setting loader`.

Author: This will be done after merging PR with changes in the code.

####################

Use the Settings load module to trigger loading the data from non-volatile memory.

Module Events
*************

+----------------+-------------+-------------------+-----------------+------------------+
| Source Module  | Input Event | This Module       | Output Event    | Sink Module      |
+================+=============+===================+=================+==================+
|                |             | ``settings_load`` |                 |                  |
+----------------+-------------+-------------------+-----------------+------------------+

Configuration
*************

> The Settings load module is enabled for every nRF Desktop device with :ref:`zephyr:settings` enabled - when :option:`CONFIG_SETTINGS` is set.

Author: The CONFIG_SETTINGS option is used to enable the zephyr:settings. I will change this a bit.

> Zephyr assumes that the application is responsible for calling the :cpp:func:`settings_load` after having done all necessary initialization.

Reviewer: ```suggestion
Zephyr assumes that the application will call :cpp:func:`settings_load` after completing all necessary initialization.
```
Any reason we can provide for this behavior? This `responsible` feels a bit off here, suggesting some kind of imperative, obligation, or requirement.

Author: I will add more detailed description here.

The mentioned function is called on the ``settings_load`` module initialization.

> Make sure that all the settings handlers have been registered and the :cpp:func:`bt_enable` was called before the ``settings_load`` module is initialized.

Reviewer: ```suggestion
Make sure that all settings handlers are registered and :cpp:func:`bt_enable` is called before the ``settings_load`` module is initialized.
```

> You can set the ``CONFIG_DESKTOP_SETTINGS_LOAD_BY_THREAD_ENABLE`` Kconfig option to load the settings by the separate thread in background instead of use of the system workqueue for that purpose.

Reviewer: Is this a separate option or something that this module allows you to do? Specify it here, please.

Author: This is separate option for the module. I will add more detailed description here.

> You can set the ``CONFIG_DESKTOP_SETTINGS_LOAD_BY_THREAD_ENABLE`` Kconfig option to load the settings by the separate thread in background instead of use of the system workqueue for that purpose.

Reviewer: ```suggestion
You can set the ``CONFIG_DESKTOP_SETTINGS_LOAD_BY_THREAD_ENABLE`` Kconfig option to load the settings in a separate thread in the background instead of using the system workqueue for that purpose.
```

> The stack size for the mentioned thread is defined as ``CONFIG_DESKTOP_SETTINGS_LOAD_THREAD_STACK_SIZE``.

Reviewer: ```suggestion
The stack size for the background thread is defined as ``CONFIG_DESKTOP_SETTINGS_LOAD_THREAD_STACK_SIZE``.
```

> Loading the data in the system workqueue context could block the mentioned workqueue and result in missing key presses on system reboot - the :ref:`buttons` module uses the system workqueue to scan the keyboard matrix.

Reviewer: ```suggestion
Loading data in the system workqueue context could block the workqueue and result in missing key presses on system reboot, given that the :ref:`buttons` module uses the system workqueue to scan the keyboard matrix.
```

> Loading the data in the system workqueue context could block the mentioned workqueue and result in missing key presses on system reboot - the :ref:`buttons` module uses the system workqueue to scan the keyboard matrix.

Reviewer: OK, so this could block the workqueque. Then why we just suggest it instead of recommending or enforcing it? This is not clear.

Author: I will add more detailed description for the problem.

> The ``CONFIG_DESKTOP_SETTINGS_LOAD_BY_THREAD_ENABLE`` is enabled for keyboard reference design (``nrf52_pca20037`` board).

Reviewer: So what impact does this have on this module? Or this is just a FYI?

.. _settings_load:

Settings load module
####################

Use the Settings load module to trigger loading the data from non-volatile memory.

Module Events
*************

+----------------+-------------+-------------------+-----------------+------------------+
| Source Module  | Input Event | This Module       | Output Event    | Sink Module      |
+================+=============+===================+=================+==================+
|                |             | ``settings_load`` |                 |                  |
+----------------+-------------+-------------------+-----------------+------------------+

Configuration
*************

The Settings load module is enabled for every nRF Desktop device with :ref:`zephyr:settings` enabled - when :option:`CONFIG_SETTINGS` is set.

> Loading all the settings in one go is more efficitent than making each subsystem load the settings independently.

Reviewer 2: This is not true. Also the fact that we load it from one place is because of how Zephyr's settings architecture is done.

I would much prefer to load settings in each module (is parts and only when needed) but it is impossible.

Author: Zephyr provides the `settings_load_subtree` function, but using it is ineffective (every time we traverse through the whole settings memory). So there is alternative solution.

Also refer to Zephyr Bluetooth settings documentation: `https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_BT_SETTINGS.html`

Author: The sentence: `Loading all the settings in one go is more efficitent than making each subsystem load the settings independently.` will be removed. It may be misleading. Discussed offline.

> Zephyr assumes that the application is responsible for calling the :cpp:func:`settings_load` after having done all necessary initialization.

Reviewer 2: It may be worth adding a list of modules we are waiting for. Or maybe we should create settnigs_load_def.h file when such list is given?

Reviewer 2: (most probably no)

Author: I think we could refer to this module from other modules that use Zephyr settings (I did it already in the Bluetooth guide)
